cfg/builder/builder_walk.cc

```diff
@@ -823,14 +823,58 @@ BasicBlock *CFGBuilder::walk(CFGContext cctx, ast::ExpressionPtr &what, BasicBlo
                 for (auto &expr : a.rescueCases) {
                     auto *rescueCase = ast::cast_tree<ast::RescueCase>(expr);
                     auto caseBody = cctx.inWhat.freshBlockWithRegion(cctx.loops, handlersRubyRegionId);
+                    auto caseBodyEntrance = caseBody;
                     auto &exceptions = rescueCase->exceptions;
                     auto added = false;
                     auto *local = ast::cast_tree<ast::Local>(rescueCase->var);
                     ENFORCE(local != nullptr, "rescue case var not a local?");
 
                     auto localVar = cctx.inWhat.enterLocal(local->localVariable);
-                    rescueHandlersBlock->exprs.emplace_back(localVar, rescueCase->var.loc(),
-                                                            make_insn<Ident>(exceptionValue));
+
+                    synthesizeExpr(caseBody, localVar, rescueCase->var.loc(), make_insn<Ident>(exceptionValue));
+
+                    // If the ensure block is present we want to propagate the exception type into it
+                    // To achieve that, we insert a BB before the rescue block. The new block branches on a `T.anything`
+                    // in a way that a true branch goes to the regular rescue block and a false branch is going to the
+                    // ensure block. That way we can model the exception being reraised in the rescue
+                    //
+                    //                       With ensure
+                    //
+                    //                   ┌──────────────────┐
+                    //                   │                  │
+                    //                   │ rescue header bb ├───┐
+                    //                   │                  │   │
+                    //  Without ensure   └────────┬─────────┘   │
+                    //                            │             │
+                    //   ┌───────────┐      ┌─────▼─────┐       │
+                    //   │           │      │           │       │
+                    //   │ rescue bb │      │ rescue bb │       │
+                    //   │           │      │           │       │
+                    //   └─────┬─────┘      └─────┬─────┘       │
+                    //         │                  │             │
+                    //         │                  │             │
+                    //   ┌─────▼─────┐      ┌─────▼─────┐       │
+                    //   │           │      │           │       │
+                    //   │ ensure bb │      │ ensure bb ◄───────┘
+                    //   │           │      │           │
+                    //   └───────────┘      └───────────┘
+                    //
+                    // See test/testdata/cfg/ensure_exception.rb for a common use case
+                    auto shouldSplitCaseBody = !ast::isa_tree<ast::EmptyTree>(a.ensure);
+                    if (shouldSplitCaseBody) {
```

Comment:
It's important to add an extra block only if the ensure block is present. 
If we would add an extra block every time, when this program would fail to typecheck with an error saying the return type is `T.nilable(T::Boolean)`
```ruby
sig {returns(T::Boolean)}
def foo
  begin 
    true
  rescue
    false
  end
end
```